AnkiDroid/src/main/java/com/ichi2/anki/jsaddons/TgzPackageExtract.kt

```diff
@@ -0,0 +1,205 @@
+ *      <p>                                                                             *
+ *      http://www.apache.org/licenses/LICENSE-2.0                                      *
+ *      <p>                                                                             *
+ *      Unless required by applicable law or agreed to in writing, software             *
+ *      distributed under the License is distributed on an "AS IS" BASIS,               *
+ *      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.        *
+ *      See the License for the specific language governing permissions and             *
+ *      limitations under the License.                                                  *
+ ****************************************************************************************/
+
+package com.ichi2.anki.jsaddons
+
+import org.apache.commons.compress.archivers.ArchiveException
+import org.apache.commons.compress.archivers.ArchiveStreamFactory
+import org.apache.commons.compress.archivers.tar.TarArchiveEntry
+import org.apache.commons.compress.archivers.tar.TarArchiveInputStream
+import org.apache.commons.compress.utils.IOUtils
+import timber.log.Timber
+import java.io.*
+import java.util.zip.GZIPInputStream
+
+/**
+* In JS Addons the addon packages are downloaded from npm registry.
+* The file format of downloaded file is tgz, so I created this class to extract tgz file to provided directory.
+*
+* This file in kotlin is used for extracting tgz file.
+* https://android.googlesource.com/platform/tools/tradefederation/+/master/src/com/android/tradefed/util/TarUtil.java
+*
+* Also added zip slip safety
+* https://snyk.io/research/zip-slip-vulnerability
+*/
+
+object TgzPackageExtract {
+    private val GZIP_SIGNATURE = byteArrayOf(0x1f, 0x8b.toByte())
+
+    /**
+     * Determine whether a file is a gzip.
+     *
+     * @param file the file to check.
+     * @return whether the file is a gzip.
+     * @throws IOException if the file could not be read.
+     */
+    @Throws(IOException::class)
+    fun isGzip(file: File?): Boolean {
+        val signature = ByteArray(GZIP_SIGNATURE.size)
+        FileInputStream(file).use { stream ->
+            if (stream.read(signature) != signature.size) {
+                return false
+            }
+        }
+        return GZIP_SIGNATURE.contentEquals(signature)
+    }
+
+    /**
+     * Untar and ungzip a tar.gz file to a AnkiDroid/addons directory.
+     *
+     * @param tarballFile the .tgz file to extract
+     * @param addonsDir   the AnkiDroid addons directory
+     * @return the temp directory.
+     * @throws FileNotFoundException if .tgz file or ungzipped file i.e. .tar file not found
+     * @throws IOException
+     */
+    @Throws(Exception::class)
+    fun extractTarGzipToAddonFolder(tarballFile: File, addonsDir: File) {
+        require(isGzip(tarballFile)) { String.format("%s is not a valid .tgz file.", tarballFile.absolutePath) }
```

Comment:
Just a tiny wording change to be more user-friendly, but I think it should be in a string constant as I think these will be exposed as errors to the user eventually?

```suggestion
        require(isGzip(tarballFile)) { String.format("%s is not a valid javascript addon package.", tarballFile.absolutePath) }
```